File: sync.py
import logging
import numpy as np
import time
import math as m
import sys
from cmath import phase
from util import flatten

from plotter import Plot

from audio import *

PI2 = m.pi * 2

def genSync(base, numSyncChans, syncChannelGap, syncChipSamples):
  syncPairs = numSyncChans / 2
  pattern = [1,0] * syncPairs
  opposite = [1-b for b in pattern]

  # combine a decent chunk or we get a buffer underrun.
  def createSyncSignal(chipsPerPulse, repeats):
    p = pattern

    def createPulse(pattern):
      return list(buildWaveform(pattern, base, syncChannelGap, chipsPerPulse * syncChipSamples))
      return list(window(buildWaveform(pattern, base, syncChannelGap, chipsPerPulse * syncChipSamples)))

    p1 = createPulse(pattern)
    p2 = createPulse(opposite)

    return (p1 + p2) * repeats + p1

  chipsPerLongPulse = 2
  metaSignalBucket = 2
  syncLong = createSyncSignal(chipsPerLongPulse, metaSignalBucket + 1)
  syncReady = createSyncSignal(1, chipsPerLongPulse * 2) # * chipsPerLongPulse * metaSignalBucket

  #fadein(syncLong, chipSamples / 20)
  #fadeout(syncReady, chipSamples / 20)

  # silence is not actually necessary, it's just to prevent
  # buffer underruns with the audio output since we're not
  # sending any data here
  # FIXME this shouldn't be necessary
  #silenc = list(silence(chipSamples * 3))

  silenc = []
  silenc = list(silence(10000))

  #ttt =  time.time()
  #time.sleep(2)
 # for i in xrange(10)
  return np.array(silenc + syncLong + syncReady )
  #print time.time() - ttt


def vectorDetector(bucketVals, bitpattern):
  assert len(bucketVals) == len(bitpattern), '%s vs %s' % (bucketVals, bitpattern)
  bitpattern = bitpattern / np.linalg.norm(bitpattern)
  bucketVals = np.abs(bucketVals)
  norm = np.linalg.norm(bucketVals)
  bucketVals = bucketVals / norm
  result = np.dot(bucketVals, bitpattern)

# hack, apparently sometimes '1.0 <= 1.0' is false,
# haven't quite figured out why...
  result *= 0.99
  result += 0.005

  result = result * 2 - 1

  return result

class SyncUtil(object):
  def __init__(self, baseBucket, spacing, numchans,
      chipsPerSyncPulse=2, numCyclesAsReadyPulses=1,
      signalFactor=1.0,
      detectStrategy=vectorDetector,
      detectionSamplesPerChip = 4,
      misalignmentTolerance = 0.15,
      verbose = False
      ):
    """sync utility for sync signals of form SyncPulse ++ ChunkPulse, e.g.
    AAABBBAAABBBAAABBBABABABAB - the groups of 3 are the sync pulses, and then
    the groups of 1 are the "about to start data" pulses, of which there are
    a fixed number.

    signalFactor: arbitrary number as threshold of signal detection; zero treats
        everything as a signal, higher numbers are increasingly selective,
        infinity would be for a perfect, pure signal

    chipsPerSyncPulse: number of chips per "up" or "down" part of sync signal.
        e.g. if the sync signal is AAABBBAAABBB where each letter has duration
        of one chip,t hen chipsPerSyncPulse is 3.

    numCyclesAsReadyPulses: duration, in sync-pulse cycles (e.g. a full AAABBB)
        of the ABAB pulses in quick succession that occur to signal the start
        of the data section. E.g. if it is 1, and chipsPerSyncPulse is 3,
        then the 'ready' signal would be ABABAB (6 pulses) before data begins.

    detectionSamplesPerChip: number of times to apply the detection window
        per signal pulse. higher means more accurate but more CPU.

    misalignmentTolerance: proportion chipSize in misalignment that is tolerated
    """
    #assert chipsPerSyncPulse >= 2
    assert numCyclesAsReadyPulses == 1 # handling other values not implemented
    self.baseBucket = baseBucket
    self.spacing = spacing
    self.numchans = numchans
    self.chipsPerSyncPulse = chipsPerSyncPulse
    self.signalFactor = signalFactor
    self.detectStrategy = detectStrategy
    self.detectionSamplesPerChip = detectionSamplesPerChip
    self.misalignmentTolerance = misalignmentTolerance
    self.numCyclesAsReadyPulses = numCyclesAsReadyPulses
    self.verbose = verbose

    self.doPlot = True # todo: flag
    self.doPlot = False# True # todo: flag
    self.plotter = None

  def align(self, receiver, chipSize, patternUnit = [1,0]):
    """Listens for a sync signal from the receiver and aligns to it.

    Blocks, and when it returns, the receiver will be aligned to the start
    of the data to be read.
    """
    assert self.numchans % len(patternUnit) == 0
    pattern = np.array(patternUnit * (self.numchans / len(patternUnit)))
    assert len(pattern) == self.numchans

    pulseSize = self.chipsPerSyncPulse * chipSize
    pulseWindow = pulseSize

    signalCycleSize = 2 * pulseSize
    metaSignalBucket = 2 # TODO: fix code to avoid redundant work
    readyMetaSignalBucket = metaSignalBucket * self.chipsPerSyncPulse

    longSignalWindow = (1 + metaSignalBucket * 2) * pulseSize
    dataWindow = longSignalWindow + chipSize

    readySize = (1 + 2 * self.chipsPerSyncPulse) * chipSize
    assert readySize == signalCycleSize + chipSize

    metaSamplesPerPulse = self.chipsPerSyncPulse * self.detectionSamplesPerChip
    metaSamplesPerCycle = 2 * metaSamplesPerPulse
    samplesPerMetaSample = float(chipSize) / self.detectionSamplesPerChip

    assert self.chipsPerSyncPulse == (pulseWindow / chipSize)
    #bucketIndices = [
    #  self.chipsPerSyncPulse * (self.baseBucket + i * self.spacing)
    #  for i in xrange(self.numchans)]
    bucketIndices = [
      (self.baseBucket + i * self.spacing)
      for i in xrange(self.numchans)]

    data = list(receiver.receiveBlock(dataWindow))

    # -1: nothing
    #  0: got signal, alignment attempted for next cycle
    #  1: got signal and aligned
    #  2: expecting 'ready for data' cycle
    state = -1
    confidence2 = -1

    tries = 0
    oldMetaSignal = None

    while True:
      shortMetaSignal = []

      for i in xrange(longSignalWindow * self.detectionSamplesPerChip / chipSize):
        start = int(i * samplesPerMetaSample)
        end = start + chipSize # + pulseWindow
        assert end <= len(data)
        spectrum = np.fft.rfft(data[start:end])
        bucketVals = [spectrum[b] for b in bucketIndices]
        match = self.detectStrategy(bucketVals, pattern)
        shortMetaSignal.append(match)

      longMetaSignal = []
      for i in xrange(metaSamplesPerCycle * metaSignalBucket):
        avg = np.mean(shortMetaSignal[i:i+metaSamplesPerPulse])
        lms = avg
       # assert avg >= -1.0 and avg <= 1.0, avg
       # lms = 1 - m.acos(avg) * 2 / m.pi
        assert lms >= -1 and lms <= 1, lms
        longMetaSignal.append(lms)

      longMetaSpectrum = np.fft.rfft(longMetaSignal)
      partALength = metaSamplesPerCycle * (metaSignalBucket - 1)
      assert partALength >= metaSamplesPerCycle
      longMetaSpectrumA = np.fft.rfft(longMetaSignal[:partALength])
      longMetaSpectrumB = np.fft.rfft(longMetaSignal[partALength:])

      partBLength = self.chipsPerSyncPulse * (self.detectionSamplesPerChip * 2)
      #shortMetaSpectrumA = np.fft.rfft(shortMetaSignal[:-partBLength])
      shortMetaSpectrumB = np.fft.rfft(shortMetaSignal[-partBLength:])

      if self.doPlot:
        def fixSpectrum(d):
          #return abs(np.array([0] + list(d[1:])))
          return abs(d)
        spectrumPlotHeight = 10# 40
        plotData = [
          (-0.5, 0.5, data),
          (-1.0, 1.0, shortMetaSignal),
          (-1.0, 1.0, shortMetaSignal[-partBLength:]),
          (0, spectrumPlotHeight, fixSpectrum(shortMetaSpectrumB)),
          (-1.0, 1.0, longMetaSignal),
          (0, spectrumPlotHeight, fixSpectrum(longMetaSpectrum)),
          (-1.0, 1.0, longMetaSignal[:partALength]),
          (0, spectrumPlotHeight, fixSpectrum(longMetaSpectrumA)),
          (-1.0, 1.0, longMetaSignal[partALength:]),
          (0, spectrumPlotHeight, fixSpectrum(longMetaSpectrumB)),
          ##(0, spectrumPlotHeight, fixSpectrum(shortMetaSpectrumA)),
          ]
        if not self.plotter:
          self.plotter = Plot([(len(d), yMin, yMax) for (yMin, yMax, d) in plotData])
        for i, (_,_ , d) in enumerate(plotData):
          self.plotter.plot(i, d)

      def largestOther(spectrum, notBucket):
        return max([
            abs(spectrum[b]) for b in xrange(len(spectrum))
            if b != 0 and b != notBucket])

      def getAlignment(spectrum, bucket, state, chips, label, dbg=False):
        if not label: label = bucket == metaSignalBucket and '.' or '!' # for logging
        bucketValue = spectrum[bucket]
        resultAmplitude = abs(bucketValue)
        largestRemaining = largestOther(spectrum, bucket)
        #logging.debug('REMAINING: %s' % largestRemaining)
        misalignment = 0

        #logging.debug('%s:%s' % (phase(bucketValue), abs(bucketValue)))
        if (abs(bucketValue) > self.signalFactor * largestRemaining):
          #logging.debug('%s, %s' % (abs(bucketValue), largestRemaining))
          misalignment = phase(bucketValue) / PI2
          assert misalignment >= -0.5 and misalignment <= 0.5
          misalignmentPerChip = misalignment * chips
          if state == -1:
            logging.debug('[%s]' % label)
            result = 0 # don't consider misalignment on first go
          elif abs(misalignmentPerChip) <= self.misalignmentTolerance:
            result = 1
            if dbg: 
              logging.debug('%s:%s' % (phase(bucketValue), abs(bucketValue)))
            logging.debug('(%s %.3f)' % (label, abs(misalignmentPerChip)))
          else:
            #logging.debug('<%s>' % label, False)
            logging.debug('<%s:%s>' % (label, misalignmentPerChip))
            #logging.debug('MISALIGNED %s %.2f' % (label, misalignmentPerChip))
            result = 0 # couldn't have been a good signal
        else:
          result = -1
          logging.debug('%s' % label)
          if dbg: 
            logging.debug('%s:%s' % (phase(bucketValue), abs(bucketValue)))

        return (result, misalignment)


      #if state == 2:
      #  #logging.debug(metaSignal)
      #  #if newState == -1:
      #  #  tries += 1
      #  #  if tries < 3:
      #  #    newState = 1
      #  #  else:
      #  #    tries = 0

      #  # check for, and verify alignment to, the ready signal
      #  #logging.debug('A')
      #  (result, dummy) = getAlignment(shortMetaSpectrum,
      #      readyMetaSignalBucket, 1, 2*chipSize, True)
      #  #print 'THIS: ', readyMetaSignalBucket
      #  if result == 1:
      #    # Got ready signal, aligned and ready!
      #    logging.debug('ALIGNED')
      #    return

      #  # Fail. start again.
      #  state = -1
      #  #import time
      #  #time.sleep(10)

      if state == 2:
        (resultShortB, dummy) = getAlignment(shortMetaSpectrumB,
            self.chipsPerSyncPulse, 1, 1, '#')
        if resultShortB == 1:
          logging.debug('ALIGNED')
          #time.sleep(5)
          #receiver.receiveBlock(int(chipSize*2.0))
          #receiver.receiveBlock(int(chipSize*0.0))
          logging.info("aligned")
          return
        else:
          #time.sleep(10)
          state = -1

      if state == 1:
        (resultLongA, dummy) = getAlignment(longMetaSpectrumA,
            metaSignalBucket - 1, 1, self.chipsPerSyncPulse, '.')
        (resultLongB, dummy) = getAlignment(longMetaSpectrumB,
            1, 1, self.chipsPerSyncPulse, '.')
        (resultShortB, chipMisalignment) = getAlignment(shortMetaSpectrumB,
            self.chipsPerSyncPulse, 1, 1, '#')
        logging.debug('\n|| %s %s %s' % (resultLongA, resultLongB, resultShortB))
        if resultLongA == 1 and resultLongB < 1 and resultShortB >= 0:
          logging.debug('========= %s %s %s' % (resultLongA, resultLongB, resultShortB))
          state = 2
          misalignment = chipMisalignment / self.chipsPerSyncPulse
          #time.sleep(10)
          #logging.debug('ALIGNED')
          #return

      if state < 2:
        # Check for, and align to, the long signal
        #logging.debug('L')
        (state, misalignment) = getAlignment(longMetaSpectrum,
            metaSignalBucket, state, self.chipsPerSyncPulse, '.')

      if state < 0:
        misalignment = 0

      if True or state < 2:
        alignedAmount = int(signalCycleSize * (1 - misalignment))
      else:
        alignedAmount = signalCycleSize * (metaSignalBucket - 1) + pulseSize
        logging.debug('aligning +%d cycles' % (metaSignalBucket - 1))

      data = data[alignedAmount:] + list(receiver.receiveBlock(alignedAmount))

File: native/sync.cpp
#include "sync.h"

#include "config.h"
#include "constants.h"
#include "spectrum.h"
#include "audio.h"
#include "log.h"

#include <iostream>
#include <cmath>
#include <cassert>

using namespace std;

static const int MAX_BUFFER_SAMPLES = SAMPLE_RATE * 10; // num in seconds

//void printArray(float* values, int length) {
//  cerr << "\n[";
//  for (int i = 0; i < length; i++) {
//    if (i > 0) cerr << ',';
//    cerr << values[i];
//  }
//  cerr << "]\n";
//}

Sync::Sync(SyncConfig* cfg, Stream<float>& source) : source(source), cfg(cfg) {
  samplesPerMetaSample = (float) cfg->chipSize / cfg->detectionSamplesPerChip;

  assert(cfg->numChannels % 2 == 0);
  bitPatternAbs = sqrt((float)cfg->numChannels/2);

  metaSamplesPerPulse = cfg->detectionSamplesPerChip * cfg->chipsPerSyncPulse;
  //cerr << metaSamplesPerPulse << "<------------\n";
  metaSamplesPerCycle = 2 * metaSamplesPerPulse;
  readyRepeats = 2 * cfg->chipsPerSyncPulse;
  pulseSamples = cfg->chipsPerSyncPulse * cfg->chipSize;

  precomp = new complex<float>*[cfg->numChannels];
  for (int i = 0; i < cfg->numChannels; i++) {
    precomp[i] = new complex<float>[cfg->chipSize];
    int bucket = cfg->baseBucket + (i * cfg->channelSpacing);
    for (int j = 0; j < cfg->chipSize; j++) {
      double arg = M_PI * 2 * j * bucket / cfg->chipSize;
      precomp[i][j] = complex<float>(cos(arg), sin(arg));
    }
  }

  reset(false);

  //cerr << "VALS " << cfg->baseBucket << ' '
      //<< cfg->numChannels << ' ' << cfg->channelSpacing << '\n';
}

Sync::~Sync() {
  for (int i = 0; i < cfg->numChannels; i++) {
    delete precomp[i];
  }
  delete precomp;
}

int Sync::getState() {
  return state;
}

void Sync::createSyncCycles(int chipsPerPulse, int cycles, vector<float> &target) {

  int n = 0;
  for (int i = 0; i < cycles * 2 + 1; i++) {
    int pattern = i % 2;
    for (int t = 0; t < cfg->chipSize * chipsPerPulse; t++) {
      float val = 0;
      for (int chan = pattern; chan < cfg->numChannels; chan += 2) {
        val += precomp[chan][t % cfg->chipSize].imag();
      }
      n++;
      target.push_back(val);
    }
  }

  normalize(target.end() - n, target.end());
  fadein(target.end() - n, cfg->chipSize / 20);
  fadeout(target.end() - cfg->chipSize / 20, cfg->chipSize / 20);
}

void Sync::generateSync(vector<float> &target) {
  createSyncCycles(cfg->chipsPerSyncPulse, cfg->longMetaBucket + 1, target);
  createSyncCycles(1, cfg->chipsPerSyncPulse * 2, target);
}

void Sync::reset(bool consume) {
  fftSampleIndex = 0;
  syncOffset = 0;
  state = -1;
  // TODO: reserve certain size in buffers.
  shortMetaBuffer.clear();
  metaBuffer.clear();

  if (consume) {
    source.consumeAll();
  }
}

bool Sync::sync() {
  if (state == -1 && source.size() > MAX_BUFFER_SAMPLES) {
    source.consumeAll();
  }

  source.pull();

  //cerr << "do ";
  //cerr << "\n\n\n===========\nCALLED with " << samples.size() << " samples\n";
  // Append samples to buffer
  // TODO: just use a rotating buffer of a fixed float array

  // need extra chipSize at the end for possible alignment signal,
  // which is 1 long cycle + 1 chip in size (sync signals need to
  // have an odd number of pulses).

  while (source.size() >= bufferStart() + pulseSamples + cfg->chipSize) {
   //// Direct version, faster for numChannels less than about 24ish
    int numChans = cfg->numChannels;
    vector<complex<float> > bucketVals(numChans);
    assert (source.size() > bufferStart() + cfg->chipSize);
    for (int c = 0; c < numChans; c++) {
      bucketVals[c] = 0;
      float *b, *bEnd = source.raw() + bufferStart() + cfg->chipSize;
      complex<float> *p, *pEnd = precomp[c] + cfg->chipSize;
      for (b = source.raw() + bufferStart(), p = precomp[c];
          b < bEnd && p < pEnd; b++, p++) {
        bucketVals[c] += (*b) * (*p);
      }
    }
    float match = detectMatch(bucketVals);
    shortMetaBuffer.push_back(match);
    fftSampleIndex++;

   //// (FFT version)
   // Spectrum spectrum(buffer.data() + bufferStart(), SAMPLE_RATE, cfg->chipSize);
   // complex<float> bucketVals[cfg->numChannels];
   // //copyBucketVals(spectrum, cfg->chipsPerSyncPulse, bucketVals);
   // copyBucketVals(spectrum, 1, bucketVals);

   // float match = detectMatch(bucketVals);
   // shortMetaBuffer.push_back(match);
   // fftSampleIndex++;

    if (fftSampleIndex >= cfg->detectionSamplesPerChip) {
      //metaBuffer.push_back(match);
      float tot = 0;
      for (int j = fftSampleIndex - cfg->detectionSamplesPerChip; j < fftSampleIndex; j++) {
        tot += shortMetaBuffer[j];
      }
      metaBuffer.push_back(tot / cfg->detectionSamplesPerChip);
    }
  }

      //return samples.end();

  //cerr << "\nsyncOffset " << syncOffset << ' ' << 2*pulseSamples << '\n';
  int longMetaSpectrumLength = metaSamplesPerCycle * cfg->longMetaBucket;

  while (true) {
    int metaBufferStart = (int) (syncOffset / samplesPerMetaSample);
    //cerr << "\n---------loop " << syncOffset << ", " << metaBufferStart << '\n';
    //assert(metaBufferStart < metaBuffer.size());

    if (metaBuffer.size() < metaBufferStart + longMetaSpectrumLength) {
      // No sync yet
      return false;
    }

    
    float *longMetaSignal = metaBuffer.data() + metaBufferStart;
    //printArray(metaBuffer.data(), metaBuffer.size());
    Spectrum longMetaSpectrum(longMetaSignal, -1, longMetaSpectrumLength);
    int longPartALength = (cfg->longMetaBucket - 1) * metaSamplesPerCycle;
    int longPartBLength = longMetaSpectrumLength - longPartALength;
    assert(longPartBLength > 0);
    assert(longPartALength >= longPartBLength);
    assert(longPartALength + longPartBLength == longMetaSpectrumLength);
    assert(longPartALength >= metaSamplesPerCycle);
    Spectrum longMetaSpectrumA(longMetaSignal, -1, longPartALength);
    Spectrum longMetaSpectrumB(longMetaSignal + longPartALength, -1, longPartBLength);

    //cerr << "LEN " << longPartALength << ' ' << longPartBLength << ' ' << longMetaSpectrumLength << '\n';

    float misalignment;

    // final ready pulse is 2 long cycles in size + 1 chip
    if (state >= 1) {
      //cerr << "X ";

      int readySignalStart = 
        syncOffset + ((cfg->longMetaBucket - 1) * 2 + 1) * pulseSamples;
      //cerr << "SYNC OFFSET " << syncOffset << ' ' << readySignalStart << '\n';
      int shortMetaSpectrumLength = metaSamplesPerCycle;
      float readyBuffer[shortMetaSpectrumLength];
      for (int i = 0; i < shortMetaSpectrumLength; i++) {
        //buffer.size() >= bufferStart() + cfg->chipSize) {
        Spectrum spectrum(
          source.raw() + readySignalStart + (int)(samplesPerMetaSample*i),
          SAMPLE_RATE, cfg->chipSize);
        vector<complex<float> > bucketVals(cfg->numChannels);

        copyBucketVals(spectrum, 1, bucketVals);
        readyBuffer[i] = detectMatch(bucketVals);
      }
      Spectrum shortMetaSpectrumB(readyBuffer, -1, shortMetaSpectrumLength);
      //printArray(readyBuffer, shortMetaSpectrumLength);
      //printArray(shortMetaBuffer.data(), shortMetaBuffer.size());
      //printArray(metaBuffer.data(), metaBuffer.size());

      if (state == 2) {
        int resultShortB = getAlignment(shortMetaSpectrumB, cfg->chipsPerSyncPulse,
            1, 1, NULL);
        if (resultShortB == 1) {
          //logging.debug('state 2 ALIGNED')
          // XXX return actual alignment
          int offset = readySignalStart + (int)(samplesPerMetaSample*shortMetaSpectrumLength) + cfg->chipSize;
          if (offset < 0) {
            ll(LOG_WARN, "SCOM", "input offset before sample start, probably bad sync");
            reset(true);
            return false;
          }
          if (offset > source.size()) {
            ll(LOG_WARN, "SCOM", "input offset after sample end, probably bad sync");
            reset(true);
            return false;
          }
          cerr << "\n\nSUCCESS " << offset << "\n\n";
          ll(LOG_INFO, "SCOM", "SUCCESS %d", offset);
          reset(false);
          source.consume(offset);
          return true;
        } else {
          state = -1;
        }
      }
      
      if (state == 1) {
        int resultLongA = getAlignment(longMetaSpectrumA,
            cfg->longMetaBucket - 1, 1, cfg->chipsPerSyncPulse, NULL);
        int resultLongB = getAlignment(longMetaSpectrumB,
            1, 1, cfg->chipsPerSyncPulse, NULL);
        float chipMisalignment;
        int resultShortB = getAlignment(shortMetaSpectrumB,
            cfg->chipsPerSyncPulse, 1, 1, &chipMisalignment);
        //logging.debug('\n|| %s %s %s' % (resultLongA, resultLongB, resultShortB))
        if (resultLongA == 1 && resultLongB < 1 && resultShortB >= 0) {
          //logging.debug('========= %s %s %s' % (resultLongA, resultLongB, resultShortB))
          //logging.debug('state 1 ALIGNED')
          state = 2;
          misalignment = chipMisalignment / cfg->chipsPerSyncPulse;
          //cerr << "STATE 2 " << misalignment << '\n';
        }
        //cerr << "XXstate " << state << ',' << resultLongA << ' ' << resultLongB 
            //<< ' ' << resultShortB << '\n';
      }
    }

    if (state < 2) {
      // Check for, and align to, the long signal
      //#logging.debug('L')
      state = getAlignment(longMetaSpectrum,
          cfg->longMetaBucket, state, cfg->chipsPerSyncPulse, &misalignment);
      //cerr << "newstate " << state << ' ' << misalignment << '\n';
    }
        //cerr << "misalignmetn " << misalignment << '\n';

    if (state < 0) {
      misalignment = 0;
    }

    //if True or state < 2:
    //  alignedAmount = int(signalCycleSize * (1 - misalignment))
    //else:
    //  alignedAmount = signalCycleSize * (metaSignalBucket - 1) + pulseSize
    //  logging.debug('aligning +%d cycles' % (metaSignalBucket - 1))

    int alignedAmount = (int)(2*pulseSamples * (1 - misalignment));

    syncOffset += alignedAmount;

    //cerr << state << ' ';

  }
}

float largestOther(Spectrum &spectrum, int notBucket) {
  // skip bucket 0 and 'notBucket'
  float max = 0;
  for (int i = 1; i < spectrum.size(); i++) {
    if (i != notBucket) {
      float v = abs(spectrum.at(i));
      if (v > max) max = v;
    }
  }
  return max;
}

#define DBG true
#define PI2 6.28318530718f

int Sync::getAlignment(Spectrum &spectrum, int bucket, int state, int numChips,
    float *misalignmentOut) {
  char label = numChips == 1 ? '!' : '.'; // for debugging
  
  complex<float> bucketValue = spectrum.at(bucket);
  float resultAmplitude = abs(bucketValue);
  float largestRemaining = largestOther(spectrum, bucket);
  //logging.debug('REMAINING: %s' % largestRemaining)
  float misalignment = 0;
  int result;

  //logging.debug('%s:%s' % (phase(bucketValue), abs(bucketValue)))
  if (abs(bucketValue) > cfg->signalFactor * largestRemaining) {
    //logging.debug('%s, %s' % (abs(bucketValue), largestRemaining))
    misalignment = arg(bucketValue) / PI2;
    assert(misalignment >= -0.5 && misalignment <= 0.5);
    float misalignmentPerChip = misalignment * numChips;
    if (state == -1) {
      //logging.debug('[%s]' % label)
      result = 0; // # don't consider misalignment on first go
    } else if (abs(misalignmentPerChip) <= cfg->misalignmentTolerance) {
      result = 1;
      if (DBG) {
        //logging.debug('%s:%s' % (phase(bucketValue), abs(bucketValue)))
      }
      //logging.debug('(%s %.3f)' % (label, abs(misalignmentPerChip)))
    } else {
      //#logging.debug('<%s>' % label, False)
      //logging.debug('<%s:%s>' % (label, misalignmentPerChip))
      //#logging.debug('MISALIGNED %s %.2f' % (label, misalignmentPerChip))
      result = 0; // couldn't have been a good signal
    }
  } else {
    result = -1;
    //logging.debug('%s' % label)
    if (DBG) {
//      logging.debug('%s:%s' % (phase(bucketValue), abs(bucketValue)))
    }
  }

  if (misalignmentOut) *misalignmentOut = misalignment;
  return result;
}


int Sync::bufferStart() {
  return (int) (samplesPerMetaSample * fftSampleIndex);
}


const float HALF_PI = 1.57079632679f;
float Sync::detectMatch(vector<complex<float> > &bucketVals) {
  // Computes dot-product of normalized bucketVals with normalized bit pattern.

  float bucketSum = 0.0;
  float result = 0;

  for (int i = 0; i < cfg->numChannels; i++) {
    float a = abs(bucketVals[i]);
    // Keep track of sum of squares, to normalise at end
    bucketSum += a*a;

    // Sync match bit pattern is 1,0,1,0,1,0...
    // Therefore add every 2nd one
    if (i % 2 == 0) {
      result += a;
    }
  }
  // Normalize
  result /= sqrt(bucketSum);
  result /= bitPatternAbs;

  // can rarely get NaNs, guard against
  result *= 0.99999;
  result += 0.000005;

  // Convert to linear correlation
  result = HALF_PI - 2*acos(result);

  // Map to range -1,1, with sinewave correlation
  result = sin(result);
  //result = result / 1.57079632679f;

  // TODO: If fed random data, this actually yields
  // an average of about -0.15, figure out why
  // (expected, zero. if fed predictable data, then
  // average can be made easily to be 0, 1, -1, as expected).
  // Not sure if it happens with random bucket values,
  // or if it's random time domain data that causes some
  // bias after FFT.


//  if (abs(result) > 0.7) {
//    cerr << "              ";
//  }
//    cerr << result << '\n';
//  cerr << result << '\n';
  if (!(result >= -1.0 && result <= 1.0)) {
    //cerr << "Got result " << result << '\n';
  }
  assert(result >= -1.0 && result <= 1.0);

  return result;
}

void Sync::copyBucketVals(Spectrum &spectrum, int numChipsInSample, vector<complex<float> > &out) {
  for (int i = 0; i < cfg->numChannels; i++) {
    out[i] = spectrum.at(
        numChipsInSample * (cfg->baseBucket + (i * cfg->channelSpacing)));
  }
}
